src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java

```diff
@@ -47,14 +68,275 @@
 		return null;
 	}
+
+	/**
+	 * Searches the model for ALL occurrences, based on three main patterns:
+	 * Case 1:
+	 * n1 e1 n2*
+	 * n2* e2 n3
+	 * 
+	 * Case 2:
+	 * n1 e1 n2*
+	 * n3 e2 n2*
+	 * 
+	 * Case 3:
+	 * n2* e1 n1
+	 * n2* e2 n3
+	 * 
+	 * 
+	 * @param graph
+	 * @return
+	 */
+	public static Set<DigramOccurence> findDigramOccurrences(Model graph) {		
+		// only disjoint occurrences are added
+		Set<DigramOccurence> occurrences = new HashSet<DigramOccurence>();
+		
+		for(int i = 0; i< CASES.length; i++) {
+			QueryExecution queryExec = queryModel(graph, CASES[i]);
+			List<QuerySolution>  results = selectModel(queryExec);
+			for(QuerySolution solution: results){
+		        
+		        RDFNode n1 = solution.get("n1");
+		        RDFNode n2 = solution.get("n2");
+		        RDFNode n3 = solution.get("n3");
+		        Property e1 = ResourceFactory.createProperty(solution.get("e1").toString());
+		        Property e2 = ResourceFactory.createProperty(solution.get("e2").toString());
+		 
+		        Statement stmt1 = null;
+		        Statement stmt2 = null;
+		        switch (CASES[i]) {
+		        	case CASE_1:
+		        		stmt1 = ResourceFactory.createStatement(n1.asResource(), e1, n2);
+			        	stmt2 = ResourceFactory.createStatement(n2.asResource(), e2, n3);
+			        	break;
+					
+					case CASE_2:
+						stmt1 = ResourceFactory.createStatement(n1.asResource(), e1, n2);
+			        	stmt2 = ResourceFactory.createStatement(n3.asResource(), e2, n2);
+					break;
+					
+					case CASE_3:
+						stmt1 = ResourceFactory.createStatement(n2.asResource(), e1, n1);
+			        	stmt2 = ResourceFactory.createStatement(n2.asResource(), e2, n3);
+					break;
+
+				default:
+					break;
+				}
+				
+				if(stmt1!= null && stmt2 != null && !stmt1.equals(stmt2)) {
+					List<RDFNode> nodes = new ArrayList<RDFNode>();
+					nodes.add(n1);
+					nodes.add(n2);
+					nodes.add(n3);
+					nodes.add(e1);
+					nodes.add(e2);
+					
+					Set<RDFNode> externals = findExternals(nodes, stmt1, stmt2, graph, CASES[i]);
+//					
+//			        Digram digram = new Digram(e1, e2, getExternalIndexes(stmt1, stmt2, externals));
+//			        digrams.add(digram);
+//			        
+			        // it's only an occurrence if it has at least one external node
+					// and a maximum of 2 external nodes 
+			        if(!externals.isEmpty() && externals.size()<3) {
+			        	DigramOccurence occurrence = new DigramOccurence(stmt1, stmt2, externals);
+			        	occurrences.add(occurrence);			        	
+			        }
+				}
+		    }
+		}
+		return occurrences;
+	}
 	
-	public static Set<DigramOccurence> findDigrams(Model graph, Digram dig) {
-		// TODO Auto-generated method stub
-		return null;
+	/**
+	 * Non overlapping occurrences, priority given to the most frequent digrams
+	 * @param occurrences
+	 * @return
+	 */
+	public static Map<Digram, Set<DigramOccurence>> findNonOverOccurrences(Set <DigramOccurence> occurrences){
```

Comment:
I like the worst case O(n^2) best case O(n) greedy approach. looks good to me.
However this is heuristic, we might have to come back to check if we can get another approach implemented in the same worst case time which is not heuristic. 
Then we can compare the two approaches and check if the greedy algorithm is sufficient in most cases.

Reply:
Acutally i have one written on paper, so i can do that.